### PR TITLE
Add celery tasks for remove API endpoint

### PR DESCRIPTION
**Pull request by yashvardhannanavati**

### CODE REVIEW

File: iib/web/api_v1.py [hunk 1]
```diff
@@ -202,6 +202,17 @@ def rm_operators():
     db.session.commit()
 
     # TODO: call the celery task to remove the operator from the index image
```
Comment (mprahl) [inline iib/web/api_v1.py]: Could you remove this line?

File: iib/workers/tasks/build.py [hunk 2]
```diff
@@ -495,6 +495,114 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
     )
 
 
+@app.task
+def handle_rm_request(operators, binary_image, request_id, from_index, add_arches=None):
+    """
+    Coordinate the the work needed to build the index image with the input bundles.
```
Comment (mprahl) [inline iib/workers/tasks/build.py]: Could you update this line?

File: iib/workers/tasks/build.py [hunk 3]
```diff
@@ -495,6 +495,114 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
...
+
+    :param list operators: a list of strings representing the name of the operators to
+        remove from the index image.
+    :param str binary_image: the pull specification of the image where the opm binary gets copied
+        from.
+    :param int request_id: the ID of the IIB build request
+    :param str from_index: the pull specification of the image containing the index that the index
+        image build will be based from.
+    :param list add_arches: the list of arches to build in addition to the arches ``from_index`` is
+        currently built for; if ``from_index`` is ``None``, then this is used as the list of arches
```
Comment (mprahl) [inline iib/workers/tasks/build.py]: `from_index` can't be `None` here

File: iib/workers/tasks/build.py [hunk 4]
```diff
@@ -495,6 +495,114 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
...
+        to build the index image for
+    :raises iib.exceptions.IIBError: if the index image build fails.
+    """
+    prebuild_info = _prepare_request_for_build(binary_image, request_id, from_index, add_arches)
+
+    error_callback = failed_request_callback.s(request_id=request_id)
+    arches = prebuild_info['arches']
+    for arch in sorted(arches):
+        # The bundles are not resolved since these are stable tags, and references
```
Comment (mprahl) [inline iib/workers/tasks/build.py]: This comment is not relevant here

File: iib/workers/tasks/build.py [hunk 5]
```diff
@@ -495,6 +495,114 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
...
+        )
+
+    if not _poll_request(request_id, arches):
+        log.error('Not finishing the request since one of the underlying builds failed')
+        return
+
+    post_build_from_index_resolved = _get_resolved_image(from_index)
+    if post_build_from_index_resolved != prebuild_info['from_index_resolved']:
+        set_request_state(
+            request_id, 'failed', 'Source index_image has changed since the build started'
```
Comment (mprahl) [inline iib/workers/tasks/build.py]:
```suggestion
            request_id, 'failed', f'{from_index} has changed since the build started'
```

File: iib/workers/tasks/build.py [hunk 6]
```diff
@@ -495,6 +495,114 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
...
+            queue=f'iib_{arch}',
+            routing_key=f'iib_{arch}',
+        )
+
+    if not _poll_request(request_id, arches):
+        log.error('Not finishing the request since one of the underlying builds failed')
+        return
+
+    post_build_from_index_resolved = _get_resolved_image(from_index)
+    if post_build_from_index_resolved != prebuild_info['from_index_resolved']:
```
Comment (mprahl) [inline iib/workers/tasks/build.py]: In a separate PR or commit, could you please add this to the `add` workflow? You could probably put this in a private function that can be used in both places.

File: iib/workers/tasks/build.py [hunk 7]
```diff
@@ -495,6 +495,114 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
...
+            '--binary-image',
+            binary_image,
+            '--from-index',
+            from_index,
+            '--operators',
+            ','.join(operators),
+        ]
+
+        log.info(
+            'Generating the database file from an existing database %s with excluding'
```
Comment (mprahl) [inline iib/workers/tasks/build.py]:
```suggestion
            'Generating the database file from an existing database %s and excluding'
```

File: iib/workers/tasks/build.py [hunk 8]
```diff
@@ -392,6 +392,21 @@ def _skopeo_inspect(*args, use_creds=False):
     return json.loads(_run_cmd(cmd, exc_msg=exc_msg))
 
 
+def _validate_index_image(resolved_prebuild_from_index, unresolved_from_index):
```
Comment (mprahl) [inline iib/workers/tasks/build.py]:
```suggestion
def _verify_index_image(resolved_prebuild_from_index, unresolved_from_index):
```

File: iib/workers/tasks/build.py [hunk 9]
```diff
@@ -392,6 +392,21 @@ def _skopeo_inspect(*args, use_creds=False):
...
+    """
+    Validates if the index image has changed since IIB build started
+
+    :param str resolved_prebuild_from_index: resolved index image before starting the build
+    :param str unresolved_from_index: unresolved index image provided as API input
+    :return: a boolean indicating if the index image has changed since IIB build started
```
Comment (mprahl) [inline iib/workers/tasks/build.py]: I think it'd be more intuitive if this raised an `IIBError` exception if the from index image changed saying `The supplied from index image changed during the IIB request. Please resubmit the request.`. Then because you have `link_error=error_callback` when you call the Celery task, it will make the request as failed with the state reason as the exception.

File: iib/workers/tasks/build.py [hunk 10]
```diff
@@ -495,6 +510,109 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
     )
 
 
+@app.task
+def handle_rm_request(operators, binary_image, request_id, from_index, add_arches=None):
+    """
+    Coordinate the the work needed to remove the input operators and rebuild the index image.
```
Comment (mprahl) [inline iib/workers/tasks/build.py]:
```suggestion
    Coordinate the work needed to remove the input operators and rebuild the index image.
```

File: iib/workers/tasks/build.py [hunk 11]
```diff
@@ -495,6 +510,109 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
...
+        )
+
+    if not _poll_request(request_id, arches):
+        log.error('Not finishing the request since one of the underlying builds failed')
+        return
+
+    if not _validate_index_image(prebuild_info['from_index_resolved'], from_index):
+        set_request_state(
+            'request_id', 'failed', f'{from_index} has changed since the build started'
+        )
```
Comment (mprahl) [inline iib/workers/tasks/build.py]: You should return after this

File: iib/workers/tasks/build.py [hunk 12]
```diff
@@ -392,6 +392,22 @@ def _skopeo_inspect(*args, use_creds=False):
     return json.loads(_run_cmd(cmd, exc_msg=exc_msg))
 
 
+def _verify_index_image(resolved_prebuild_from_index, unresolved_from_index):
+    """
+    Verifies if the index image has changed since IIB build started
```
Comment (mprahl) [inline iib/workers/tasks/build.py]:
```suggestion
    Verify if the index image has changed since the IIB build request started
```

File: iib/workers/tasks/build.py [hunk 13]
```diff
@@ -495,6 +511,106 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
...
+    :param list add_arches: the list of arches to build in addition to the arches ``from_index`` is
+        currently built for.
+    :raises iib.exceptions.IIBError: if the index image build fails.
+    """
+    prebuild_info = _prepare_request_for_build(binary_image, request_id, from_index, add_arches)
+
+    error_callback = failed_request_callback.s(request_id=request_id)
+    arches = prebuild_info['arches']
+    for arch in sorted(arches):
+        # from_index is not resolved because docker does not support digest references
```
Comment (mprahl) [inline iib/workers/tasks/build.py]: Could you link `https://github.com/containers/libpod/issues/5234` here?

File: tests/test_web/test_api_v1.py [hunk 14]
```diff
@@ -372,7 +372,8 @@ def test_patch_request_success(db, worker_auth_env, client):
     assert rv_json == response_json
 
 
-def test_remove_operator_success(db, auth_env, client):
+@mock.patch('iib.web.api_v1.handle_rm_request')
```
Comment (mprahl) [inline tests/test_web/test_api_v1.py]: Could you add `mock_rm.assert_called_once()` to this test?

File: tests/test_workers/test_tasks/test_build.py [hunk 15]
```diff
@@ -360,3 +360,86 @@ def test_run_cmd_failed(mock_sub_run, caplog, exc_msg):
...
+
+
+@pytest.mark.parametrize('request_succeeded', (True, False))
+@mock.patch('iib.workers.tasks.build._prepare_request_for_build')
+@mock.patch('iib.workers.tasks.build.opm_index_rm')
+@mock.patch('iib.workers.tasks.build._poll_request')
+@mock.patch('iib.workers.tasks.build._verify_index_image')
+@mock.patch('iib.workers.tasks.build._finish_request_post_build')
+def test_handle_rm_request(
+    mock_frpb, mock_vii, mock_pr, mock_oia, mock_prfb, request_succeeded,
```
Comment (mprahl) [inline tests/test_workers/test_tasks/test_build.py]:
```suggestion
    mock_frpb, mock_vii, mock_pr, mock_oir, mock_prfb, request_succeeded,
```

File: iib/workers/tasks/build.py [hunk 16]
```diff
@@ -495,6 +511,107 @@ def opm_index_add(bundles, binary_image, request_id, from_index=None):
...
```
Comment (mprahl) [inline iib/workers/tasks/build.py]:
```suggestion
        # from_index is not resolved because podman does not support digest references when performing `podman save`
```

**Comment by yashvardhannanavati** [comment]
@mprahl I decided to add the new commit we discussed about in this PR itself. Could you PTAL?

File: iib/workers/tasks/build.py [hunk 17]
```diff
@@ -211,12 +211,17 @@ def _get_image_arches(pull_spec):
...
-        arches.add(manifest['platform']['architecture'])
+    if skopeo_raw['mediaType'] == 'application/vnd.docker.distribution.manifest.list.v2+json':
+        for manifest in skopeo_raw['manifests']:
+            arches.add(manifest['platform']['architecture'])
+    elif skopeo_raw['mediaType'] == 'application/vnd.docker.distribution.manifest.v2+json':
+        skopeo_out = _skopeo_inspect(f'docker://{pull_spec}')
+        arches.add(skopeo_out['Architecture'])
+    else:
+        raise IIBError(
+            f'The pull specification of {pull_spec} is neither a v2 manifest list nor a manifest'
```
Comment (mprahl) [inline iib/workers/tasks/build.py]:
```suggestion
            f'The pull specification of {pull_spec} is neither a v2 manifest list nor a v2 manifest'
```

File: tests/test_workers/test_tasks/test_build.py [hunk 18]
```diff
@@ -109,10 +109,20 @@ def test_get_image_arches(mock_si):
...
+    rv = build._get_image_arches('image:latest')
+    assert rv == {'amd64'}
+
+
 @mock.patch('iib.workers.tasks.build._skopeo_inspect')
 def test_get_image_arches_not_manifest_list(mock_si):
-    mock_si.return_value = {'mediaType': 'application/vnd.docker.distribution.manifest.v2+json'}
-    with pytest.raises(IIBError, match='.+is not a v2 manifest list'):
+    mock_si.return_value = {'mediaType': 'application/vnd.docker.distribution.notmanifest.v2+json'}
+    with pytest.raises(IIBError, match='.+is neither a v2 manifest list nor a manifest'):
```
Comment (mprahl) [inline tests/test_workers/test_tasks/test_build.py]:
```suggestion
    with pytest.raises(IIBError, match='.+is neither a v2 manifest list nor a v2 manifest'):
```

**Comment by mprahl** [comment]
Could you rephrase your last commit message from `Teach IIB to get image arches when it's a manifest and no manifest list` to `Teach IIB to get image arches from v2 manifests`?
